src/libvirtApi/domain.js

```diff
@@ -1112,3 +1113,23 @@ export function domainUpdateDiskAttributes({ connectionName, objPath, target, re
                 return call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'DomainDefineXML', [updatedXML], { timeout, type: 's' });
             });
 }
+
+export async function domainReplaceSpice({ connectionName, id: objPath }) {
+    /* Ideally this would be done by virt-xml, but it doesn't offer that functionality yet
+     * see https://issues.redhat.com/browse/RHEL-17436 */
+    const [domXML] = await call(connectionName, objPath, 'org.libvirt.Domain', 'GetXMLDesc', [Enum.VIR_DOMAIN_XML_INACTIVE], { timeout, type: 'u' });
+    const updatedXML = replaceSpice(domXML);
+
+    // check that updatedXML is valid; if not, throw; it needs to be updated manually
+    await (cockpit.spawn(["virt-xml-validate", "-"], { err: "message" }).input(updatedXML));
+
+    await call(connectionName, objPath, 'org.libvirt.Domain', 'Undefine', [0], { timeout, type: 'u' });
+    try {
+        await call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'DomainDefineXML', [updatedXML], { timeout, type: 's' });
+    } catch (ex) { // not-covered: unreachable, belt-and-suspenders
+        // if this fails, we are in trouble; but at least try to restore the original XML
+        console.warn("domainRplaceSpice defining updated XML", updatedXML, "failed, restoring original:", ex); // not-covered: see above
+        await call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'DomainDefineXML', [domXML], { timeout, type: 's' }); // not-covered: see above
```

Comment:
We might want to ignore errors here.

Reply:
If this fails, we are in big big trouble. We just killed the user's VM and restoring the backup also failed. Then I think it's ok for the UI to show the "migration failed" error, instead of silently pretending that it worked. But of course your VM will not exist any more either way.

Reply:
In this apocalyptic scenario, would it be possible to dump the old and new xml to a file on the host and tell the full paths to the user in the console warning? So the user can go and troubleshoot from there, either making the old one define or fixing whatever is wrong with the new one.

Reply:
This fortunately went away entirely, see the [discussion above](https://github.com/cockpit-project/cockpit-machines/pull/1387/files#r1461069400). I think this should be ok now. Thanks!